File: matrix-chain-multiplication/striver_bottom_up_dp.cpp
#include <bits/stdc++.h> 
// Time complexity: O(N*3), Space complexity: O(N^2)
int matrixMultiplication(vector<int> &arr, int N) {
    // Write your code here.
    vector<vector<int>> dp(N, vector<int>(N, 0));

    for (int st = N - 1; st >= 1; st--) {
        for (int end = st + 1; end < N; end++) {
            int minMults = 1e9;
            for (int k = st; k < end; k++) {
                int mults = arr[st - 1]*arr[k]*arr[end] + dp[st][k] + dp[k + 1][end];
                minMults = min(minMults, mults);
            }
            dp[st][end] = minMults;
        }
    }

    return dp[1][N - 1];
}

File: matrix_chain_multiplication/py_solns/striver_bottom_up_dp.py
# Time complexity: O(n^3)
# Space complexity: O(n^2)

class Solution:
    
    def init_dp_arr(self, rows, cols, init_val):
        dp = [[init_val for _ in range(cols)] for _ in range(rows)]
        return dp
        
    def matrixMultiplication(self, arr):
        # code here
        
        n = len(arr)
        
        # Base case first.
        dp = self.init_dp_arr(n, n, 0)
        
        for i in range(n-1, 0, -1):
            for j in range(i + 1, n):
                min_steps = 1e9
                for k in range(i, j):
                    # num_steps to multiply the final matrix
                    # + num_steps to multiply the 2 sub-matrices separately.
                    steps = arr[i-1]*arr[k]*arr[j] + dp[i][k] + dp[k + 1][j]
                    
                    min_steps = min(min_steps, steps)
                dp[i][j] = min_steps

        
        return dp[1][n-1]
        